Validateable object with beforeValidate() callback is not serializable

Grails 3.3.6
GORM 6.1.11.BUILD-SNAPSHOT

The `BeforeValidateHelper` implements Serializable but the `eventTriggerCallerCache` has a value of `BeforeValidateEventTriggerCaller` which is NOT Serializable.

Since it's just a cache and populates on demand, perhaps the cache could be `transient` anyway.

Attempting to serialize a command object that is `Validateable` with a `beforeValidate()` callback results in:

```
Caused by: java.io.NotSerializableException: org.grails.datastore.gorm.support.BeforeValidateHelper$BeforeValidateEventTriggerCaller
        - custom writeObject data (class "java.util.concurrent.ConcurrentHashMap")
        - object (class "java.util.concurrent.ConcurrentHashMap", {class com.triu.operational.ptl.DeclareByCommand=org.grails.datastore.gorm.support.BeforeValidateHelper$BeforeValidateEventTriggerCaller@2f9034bb})
        - field (class "org.grails.datastore.gorm.support.BeforeValidateHelper", name: "eventTriggerCallerCache", type: "interface java.util.Map")
        - object (class "org.grails.datastore.gorm.support.BeforeValidateHelper", org.grails.datastore.gorm.support.BeforeValidateHelper@48a70153)
        - field (class "com.triu.operational.ptl.DeclareByCommand", name: "grails_validation_Validateable__beforeValidateHelper", type: "class org.grails.datastore.gorm.support.BeforeValidateHelper")
```

--- a/grails-datastore-gorm-validation/src/main/groovy/org/grails/datastore/gorm/support/BeforeValidateHelper.java
+++ b/grails-datastore-gorm-validation/src/main/groovy/org/grails/datastore/gorm/support/BeforeValidateHelper.java
@@ -16,6 +16,8 @@ package org.grails.datastore.gorm.support;
 
 import groovy.lang.MetaClass;
 
+import java.io.IOException;
+import java.io.ObjectInputStream;
 import java.io.Serializable;
 import java.util.List;
 import java.util.Map;
@@ -23,7 +25,7 @@ import java.util.concurrent.ConcurrentHashMap;
 
 public class BeforeValidateHelper implements Serializable {
     public static final String BEFORE_VALIDATE = "beforeValidate";
-    private Map<Class<?>, BeforeValidateEventTriggerCaller> eventTriggerCallerCache = new ConcurrentHashMap<Class<?>, BeforeValidateEventTriggerCaller>();
+    private transient Map<Class<?>, BeforeValidateEventTriggerCaller> eventTriggerCallerCache = new ConcurrentHashMap<Class<?>, BeforeValidateEventTriggerCaller>();
     
     public static final class BeforeValidateEventTriggerCaller {
         EventTriggerCaller eventTriggerCaller;
@@ -55,5 +57,11 @@ public class BeforeValidateHelper implements Serializable {
             eventTriggerCallerCache.put(domainClass, eventTriggerCaller);
         }
         eventTriggerCaller.call(target, validatedFieldsList);
-   }
+    }
+
+    // Ensure that the cache is re-initalized empty when deserialized
+    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
+        in.defaultReadObject();
+        eventTriggerCallerCache = new ConcurrentHashMap<Class<?>, BeforeValidateEventTriggerCaller>();
+    }
 }

--- a/grails-datastore-gorm/src/test/groovy/org/grails/datastore/gorm/support/BeforeValidateHelperTests.groovy
+++ b/grails-datastore-gorm/src/test/groovy/org/grails/datastore/gorm/support/BeforeValidateHelperTests.groovy
@@ -48,6 +48,30 @@ class BeforeValidateHelperTests extends GroovyTestCase {
         assertEquals 'wrong list arg counter value', 2, obj.listArgCounter
         assertEquals 'wrong no arg counter value', 1, obj.noArgCounter
     }
+
+    void testSerialization() {
+        // Make sure something is in the cache
+        def obj = new ClassWithNoArgBeforeValidate()
+        beforeValidateHelper.invokeBeforeValidate(obj, null)
+        assertEquals 'wrong counter value', 1, obj.noArgCounter
+
+        // Serialize
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()
+        ObjectOutputStream oos = new ObjectOutputStream(byteArrayOutputStream)
+        oos.writeObject(beforeValidateHelper)
+        oos.close()
+        assertTrue 'class is serialized', byteArrayOutputStream.toByteArray().length > 0
+
+        // Deserialize
+        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(byteArrayOutputStream.toByteArray())
+        ObjectInputStream ois = new ObjectInputStream(byteArrayInputStream)
+        beforeValidateHelper = ois.readObject() as BeforeValidateHelper
+        assertNotNull 'class is deserialized', beforeValidateHelper
+
+        // Ensure it still works
+        beforeValidateHelper.invokeBeforeValidate(obj, null)
+        assertEquals 'wrong counter value', 2, obj.noArgCounter
+    }
 }
 
 class ClassWithNoArgBeforeValidate {

